lightning/src/ln/onchain_utils.rs

```diff
@@ -1,8 +1,698 @@
+	// or counterparty successful claim.
+	feerate_previous: u64,
+	// Cache of next height at which fee-bumping and rebroadcast will be attempted. In
+	// the future, we might abstract it to an observed mempool fluctuation.
+	height_timer: Option<u32>,
+	// Confirmation height of the claimed outputs set transaction. In case of reorg reaching
+	// it, we wipe out and forget the package.
+	height_original: u32,
+}
+
+impl PackageTemplate {
+	pub(crate) fn is_malleable(&self) -> bool {
+		self.malleability == PackageMalleability::Malleable
+	}
+	pub(crate) fn timelock(&self) -> u32 {
+		self.absolute_timelock
+	}
+	pub(crate) fn aggregation(&self) -> bool {
+		self.aggregation
+	}
+	pub(crate) fn get_feerate(&self) -> u64 {
+		self.feerate_previous
+	}
+	pub(crate) fn load_feerate(&mut self, new_feerate: u64) {
+		self.feerate_previous = new_feerate;
+	}
+	pub(crate) fn get_timer(&self) -> Option<u32> {
+		if let Some(ref timer) = self.height_timer {
+			return Some(*timer);
+		}
+		None
+	}
+	pub(crate) fn load_timer(&mut self, new_timer: Option<u32>) {
+		self.height_timer = new_timer;
+	}
+	pub(crate) fn outpoints(&self) -> Vec<&BitcoinOutPoint> {
+		let mut outpoints = Vec::with_capacity(self.inputs.len());
+		for (o, _) in self.inputs.iter() {
+			outpoints.push(o);
+		}
+		outpoints
+	}
+	pub(crate) fn split_package(&mut self, split_outp: &BitcoinOutPoint) -> Option<PackageTemplate> {
+		match self.malleability {
+			PackageMalleability::Malleable => {
+				let mut split_package = None;
+				let timelock = self.absolute_timelock;
+				let aggregation = self.aggregation;
+				let feerate_previous = self.feerate_previous;
+				let height_timer = self.height_timer;
+				let height_original = self.height_original;
+				self.inputs.retain(|outp| {
+					if *split_outp == outp.0 {
+						split_package = Some(PackageTemplate {
+							inputs: vec![(outp.0, outp.1.clone())],
+							malleability: PackageMalleability::Malleable,
+							absolute_timelock: timelock,
+							aggregation,
+							feerate_previous,
+							height_timer,
+							height_original,
+						});
+						return false;
+					}
+					return true;
+				});
+				return split_package;
+			},
+			_ => {
+				// Note, we may try to split on remote transaction for
+				// which we don't have a competing one (HTLC-Success before
+				// timelock expiration). This explain we don't panic!
+				// We should refactor OnchainTxHandler::block_connected to
+				// only test equality on competing claims.
+				return None;
+			}
+		}
+	}
+	pub(crate) fn merge_package(&mut self, mut merge_from: PackageTemplate) {
+		assert_eq!(self.height_original, merge_from.height_original);
+		if self.malleability == PackageMalleability::Untractable || merge_from.malleability == PackageMalleability::Untractable {
+			panic!("Merging template on untractable packages");
+		}
+		if !self.aggregation || !merge_from.aggregation {
+			panic!("Merging non aggregatable packages");
+		}
```

Comment:
Can you precise further your thinking here ? Ofc I can encode deeper aggregation/malleability at the output-level and if aggregation logic is correct it should ensure similarity of aggregation/malleability among package members. Though it is a big win in term of API clarity ? 

Further it will inflate serialized package size as you will redundantly encode properties ?

Reply:
> Can you precise further your thinking here ? Ofc I can encode deeper aggregation/malleability at the output-level (i.e `PackageSolvingData`) and if aggregation logic is correct it should ensure similarity of aggregation/malleability among package members. Though is it a big win in term of API clarity ?
> 

I'm thinking something similar to `InvoiceBuilder` where `PackageTemplate` would be constructed piecemeal (i.e., chaining method calls for each field) and where these properties are enforced by the type system by conditionally implementing methods (e.g., a `PackageTemplate` marked by a type parameter as `Malleable` could only be merged with another such `PackageTemplate`). Similarly with `aggregation`.

Reply:
TBH, I don't see a straightforward way to do such conditional methods implementation and considering the API should change a bit with remaining PRs of the patchset (e.g some output type becoming malleable), I would wait for them to land before to solidify the API now ? I think that's a good direction, just maybe too early.